File: v2/src/main/java/com/api/v2/services/impl/PersonModificationServiceImpl.java
package com.api.v2.services.impl;

import com.api.v2.domain.Person;
import com.api.v2.domain.PersonAuditTrail;
import com.api.v2.domain.PersonAuditTrailRepository;
import com.api.v2.domain.PersonRepository;
import com.api.v2.dtos.PersonModificationDto;
import com.api.v2.services.PersonModificationService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
public class PersonModificationServiceImpl implements PersonModificationService {

    private final PersonRepository personRepository;
    private final PersonAuditTrailRepository personAuditTrailRepository;

    public PersonModificationServiceImpl(
            PersonRepository personRepository,
            PersonAuditTrailRepository personAuditTrailRepository
    ) {
        this.personRepository = personRepository;
        this.personAuditTrailRepository = personAuditTrailRepository;
    }

    @Override
    public Mono<Person> modify(@NotNull Person person, @Valid PersonModificationDto modificationDto) {
        return Mono.defer(() -> {
            PersonAuditTrail auditTrail = PersonAuditTrail.create(person);
            return personAuditTrailRepository
                    .save(auditTrail)
                    .then(Mono.defer(() -> {
                        person.modify(modificationDto);
                        return personRepository.save(person);
                    }));
        });
    }
}

File: v2/src/main/kotlin/com/api/v2/people/services/PersonModificationServiceImpl.kt
package com.api.v2.people.services

import com.api.v2.people.domain.Person
import com.api.v2.people.domain.PersonAuditTrail
import com.api.v2.people.domain.PersonAuditTrailRepository
import com.api.v2.people.domain.PersonRepository
import com.api.v2.people.dtos.PersonModificationDto
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Service

@Service
internal class PersonModificationServiceImpl(
    private val personRepository: PersonRepository,
    private val personAuditTrailRepository: PersonAuditTrailRepository
): PersonModificationService {
    override suspend fun modify(person: @NotNull Person, modificationDto: @Valid PersonModificationDto): Person {
        return withContext(Dispatchers.IO) {
            val personAuditTrail = PersonAuditTrail.create(person)
            personAuditTrailRepository.save(personAuditTrail)
            person.modify(modificationDto)
            personRepository.save(person)
        }
    }
}